File: Packages/Flip54Core/Sources/Flip54Core/Card.swift
public enum Card: Codable, Hashable, Sendable {
    case standard(suit: Suit, rank: Rank)
    case joker(variant: JokerVariant)

    public static func standardDeck() -> [Card] {
        var deck: [Card] = []
        for suit in Suit.allCases {
            for rank in Rank.allCases {
                deck.append(.standard(suit: suit, rank: rank))
            }
        }
        deck.append(.joker(variant: .red))
        deck.append(.joker(variant: .black))
        return deck
    }

    /// 27-card challenge deck: drops every 2..7 and one of the 8s/jokers,
    /// keeping all face cards, Aces, and high pip values across every suit.
    /// Composition: 4×{A,K,Q,J,10,9} (24) + 2×8s (♥, ♠) + 1 joker = 27.
    public static func halfDeck() -> [Card] {
        var deck: [Card] = []
        let topRanks: [Rank] = [.ace, .king, .queen, .jack, .ten, .nine]
        for suit in Suit.allCases {
            for rank in topRanks {
                deck.append(.standard(suit: suit, rank: rank))
            }
        }
        deck.append(.standard(suit: .hearts, rank: .eight))
        deck.append(.standard(suit: .spades, rank: .eight))
        deck.append(.joker(variant: .red))
        return deck
    }

    public var isAce: Bool {
        if case .standard(_, .ace) = self { return true }
        return false
    }

    public var isJoker: Bool {
        if case .joker = self { return true }
        return false
    }

    public var isFaceCard: Bool {
        if case .standard(_, let rank) = self { return rank.isFace }
        return false
    }

    public var suit: Suit? {
        if case .standard(let suit, _) = self { return suit }
        return nil
    }
}

public enum JokerVariant: String, Codable, Hashable, Sendable {
    case red, black
}

File: android/core/src/main/kotlin/com/flip54/core/Card.kt
package com.flip54.core

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
sealed class Card {
    @Serializable
    @SerialName("standard")
    data class Standard(val suit: Suit, val rank: Rank) : Card()

    @Serializable
    @SerialName("joker")
    data class Joker(val variant: JokerVariant) : Card()

    val isAce: Boolean get() = this is Standard && rank == Rank.ACE
    val isJoker: Boolean get() = this is Joker
    val isFaceCard: Boolean get() = this is Standard && rank.isFace
    val suit: Suit? get() = (this as? Standard)?.suit

    companion object {
        fun standardDeck(): List<Card> = buildList {
            for (suit in Suit.entries) {
                for (rank in Rank.entries) {
                    add(Standard(suit, rank))
                }
            }
            add(Joker(JokerVariant.RED))
            add(Joker(JokerVariant.BLACK))
        }

        fun halfDeck(): List<Card> {
            val topRanks = listOf(Rank.ACE, Rank.KING, Rank.QUEEN, Rank.JACK, Rank.TEN, Rank.NINE)
            return buildList {
                for (suit in Suit.entries) {
                    for (rank in topRanks) {
                        add(Standard(suit, rank))
                    }
                }
                add(Standard(Suit.HEARTS, Rank.EIGHT))
                add(Standard(Suit.SPADES, Rank.EIGHT))
                add(Joker(JokerVariant.RED))
            }
        }
    }
}

@Serializable
enum class JokerVariant { RED, BLACK }
